businessLogicLayer/Features/PricesNearMe/StationDataManager.swift

```diff
@@ -0,0 +1,123 @@
+    
+    // MARK: - Constructor
+    
+    init(fuelProviders: [String: String]) {
+        // a constructor (initialiser) should only initialize its properties. It shouldn't be used to trigger network calls or any other hidden behaviour.
+        self.fuelProviders = fuelProviders
+    }
+    
+    // MARK: - Publicly available functions / features (make it easy-to-use)
+    
+    func getStationData(forPostcode postcode: String, completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        guard isRefreshingData == false else {
+            // Note: ALWAYS call your completion handlers in 100% of scenarios 😉
+            completion(.failure(.refreshAlreadyInProgress))
+            return
+        }
+        isRefreshingData = true
+        let endpointAddresses = fuelProviders.map{$1}
+        gatherStationData(from: endpointAddresses) { result in
+            DispatchQueue.main.async {
+                self.isRefreshingData = false
+                switch result {
+                case .success(let stations):
+                    self.dateOfLastRefresh = Date.now
+                    // now that we are running back on the main thread we can set the array of stations to this instance of StationDataManager. But not before. Unless we use a queue to handle read and write access.
+                    self.stationLibrary = stations // update the library of national petrol stations
+                    let filteredResults = self.filterStations(stations, forPostcode: postcode)
+                    completion(.success(filteredResults))
+                case .failure(let error):
+                    completion(result)
+                }
+            }
+        }
+    }
+    
+    // MARK: - Gathering data from StationDataAPI
+    
+    func gatherStationData(from endPointAddresses: [String], completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        let stationDataAPI = self.stationDataAPI
+        let dispatchGroup = DispatchGroup()
+        var responses: [ResponseData] = []
+        
+        for (endPointAddress) in endPointAddresses {
+            guard let url = URL(string: endPointAddress) else {
+                print("Invalid URL for \(endPointAddress)")
+                continue
+            }
+            
+            dispatchGroup.enter()
+            
+            // let the API perform the network request. It shouldn't be aware of how to handle a list of urls. Instead, we simply make multiple calls using it. It's the role of that class to perform a network request and this class to loop through many urls. 😃
+            // Also it know means that we could use the stationDataAPI to make one single call if we wanted it to - it's even more flexible now.
+            stationDataAPI.requestStationData(from: url) { result in
+                defer {
+                    dispatchGroup.leave()
+                }
+                switch result {
```

Comment:
The `Result`enum is AWESOME and really helps handling errors so easy. Take a look, I totally recommend it.